interface/src/app/sample_annotation/annotation.js

```diff
@@ -0,0 +1,117 @@
+/**
+ * "adage.sampleAnnotation" module.
+ */
+
+angular.module('adage.sampleAnnotation', [
+  'ui.router',
+  'ui.bootstrap',
+  'greenelab.utils'
+])
+
+.config(['$stateProvider', function($stateProvider) {
+  $stateProvider.state('sampleAnnotation', {
+    url: '/sample_annotation?node&samples',
+    views: {
+      main: {
+        templateUrl: 'sample_annotation/annotation.tpl.html',
+        controller: 'SampleAnnotationCtrl as ctrl'
+      }
+    },
+    data: {pageTitle: 'Sample Annotations'}
+  });
+}])
+
+.controller('SampleAnnotationCtrl', [
+  '$stateParams', '$q', '$http', '$log', 'ApiBasePath', 'errGen',
+  function($stateParams, $q, $http, $log, ApiBasePath, errGen) {
+    var self = this;
+    self.queryStatus = 'Connecting to the server ...';
+
+    // Ensure that the URL includes sample ID(s).
+    if (!$stateParams.samples) {
+      self.queryStatus = 'Please specify sample ID(s) in the URL.';
+      return;
+    }
+
+    var samplesInUrl = $stateParams.samples;
+    self.uniqueAnnotationTypes = [];
+    // Note: Ideally self.uniqueAnnotationTypes should be a hashtable (such as
+    // an object of Set) instead of Array, but Set is only available since ES6.
+
+    self.hasNode = false;
+    var sampleDict = {};
+
+    // Function that sets valid sample IDs asynchronously:
+    var getSampleIDPromise = function() {
+      var deferred = $q.defer();
+      // If 'node' is specified in the URL, query the node first:
+      if ($stateParams.node) {
+        self.hasNode = true;
+        var nodeID = $stateParams.node;
+        var activityApiConfig = {
+          params: {'node': nodeID, 'sample__in': samplesInUrl}
+        };
+        $http.get(ApiBasePath + 'activity/', activityApiConfig)
```

Comment:
Rather than making direct `$http` calls from the controller, please arrange your calls to the back end via a more abstract service layer.